src/SimulationMap.cuh

```diff
@@ -49,6 +49,72 @@ public:
     Polyhedron *const polyhedron;
 
 private:
+    /**
+     * Returns coordinates of neighbor node with or without projection on polyhedron
+     *
+     * @param current_node_id Index of the node whose neighbor is searched
+     * @param top_direction Direction vector from current node to its top neighbor
+     * @param angle Angle between the top neighbor node and the neighbor node whose index is searched
+     *              relative to current node, clockwise is positive direction
+     * @param do_projection If `true`, counted coordinates will be projected on polyhedron, otherwise they will not
+     *
+     * @returns Coordinates of neighbor node projected on polyhedron if `do_projection` is `true`,
+     *          coordinates of neighbor node without projection on polyhedron otherwise
+     */
+    __device__ SpacePoint count_neighbor_node_coordinates(int current_node_id, SpacePoint top_direction, double angle,
+                                                          bool do_projection) const;
+
+
+    /**
+     * Returns the index of the nearest node to the given point in `nodes` array
+     *
+     * Searches in the whole `nodes` array to find the nearest node, not using the pointers to neighbor nodes
+     *
+     * @param dest Coordinates of the point to find the nearest node to
+     *
+     * @note It's guaranteed that the result is the closest node to `dest`
+     *
+     * @returns The index of the found node in `nodes` array
+     */
+    __device__ int find_index_of_nearest_node(SpacePoint dest) const;
+
+
+    /**
+     * Counts direction vector from neighbor of current node to its top neighbor and sets it to `nodes_direction` array
+     *
+     * @param current_node_id Index of the node whose neighbor was searched
+     * @param neighbor_node_id Index of neighbor node
+     * @param nodes_directions Pointer to the array of direction vectors to the top neighbor node from each node
+     * @param angle Angle between the top neighbor node and the neighbor node whose index is searched
+     *              relative to current node, clockwise is positive direction
+     */
+    __device__ void set_direction_to_top_neighbor(int current_node_id, int neighbor_node_id,
+                                                  SpacePoint **nodes_directions, double angle) const;
+
+
+    /**
+     * Returns the index of nearest node to neighbor in `nodes` array, if their coordinates are almost the same
+     * or if `create_new_nodes` is `false`
+     *
+     * Creates new node in given coordinates if it does not exists and if it is possible and returns its index
+     *
+     * Returns `-1` if node cannot be created and `create_new_nodes` is `true`
+     *
+     * Node cannot be created if on the face it belongs to an another node exists and
+     * their directions to the top neighbor are not the same
```

Comment:
1. An another, omg...
2. Probably this paragraph should be moved up to achieve top-down readability
3. I think I understand this in a wrong way. If there is a node on a face, and its direction is different from mine, I can't be created? Does it mean two nodes with the same direction can't exist on the same face?.. If yes, what is the purpose/reason for this limit?